File: Bouchonnois.Tests/Unit/PrendreLApéro.cs
using Bouchonnois.Domain.Apéro;
using Bouchonnois.UseCases;
using PrendreLapéro = Bouchonnois.UseCases.PrendreLapéro;

namespace Bouchonnois.Tests.Unit
{
    public class PrendreLApéro : UseCaseTest<PrendreLapéro, VoidResponse>
    {
        public PrendreLApéro() : base((r, p) => new PrendreLapéro(r, p))
        {
        }

        [Fact]
        public void QuandLaPartieEstEnCours()
        {
            Given(
                UnePartieDeChasseExistante(
                    SurUnTerrainRicheEnGalinettes()
                )
            );

            When(id => _useCase.Handle(new Domain.Apéro.PrendreLapéro(id)));

            Then((response, partieDeChasse) =>
            {
                response.Should().Be(VoidResponse.Empty);
                partieDeChasse
                    .Should()
                    .HaveEmittedEvent(Repository, new ApéroDémarré(partieDeChasse!.Id, Now));
            });
        }

        public class Echoue : UseCaseTest<PrendreLapéro, VoidResponse>
        {
            public Echoue() : base((r, p) => new PrendreLapéro(r, p))
            {
            }

            [Fact]
            public void CarPartieNexistePas()
            {
                Given(UnePartieDeChasseInexistante());

                When(id => _useCase.Handle(new Domain.Apéro.PrendreLapéro(id)));

                ThenFailWith(
                    $"La partie de chasse {_partieDeChasseId} n'existe pas",
                    savedPartieDeChasse => savedPartieDeChasse.Should().BeNull()
                );
            }

            [Fact]
            public void SiLesChasseursSontDéjaEnApero()
            {
                Given(
                    UnePartieDeChasseExistante(
                        SurUnTerrainRicheEnGalinettes()
                            .ALapéro())
                );

                When(id => _useCase.Handle(new Domain.Apéro.PrendreLapéro(id)));

                ThenFailWith("On est déjà en plein apéro");
            }

            [Fact]
            public void SiLaPartieDeChasseEstTerminée()
            {
                Given(
                    UnePartieDeChasseExistante(
                        SurUnTerrainRicheEnGalinettes()
                            .Terminée())
                );

                When(id => _useCase.Handle(new Domain.Apéro.PrendreLapéro(id)));

                ThenFailWith("La partie de chasse est déjà terminée");
            }
        }
    }
}

File: java/src/test/java/bouchonnois/tests/unit/PrendreLApéro.java
package bouchonnois.tests.unit;

import bouchonnois.service.exceptions.LaPartieDeChasseNexistePas;
import bouchonnois.service.exceptions.OnEstDéjaEnTrainDePrendreLapéro;
import bouchonnois.service.exceptions.OnPrendPasLapéroQuandLaPartieEstTerminée;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static bouchonnois.tests.assertions.PartieDeChasseAssert.assertPartieDeChasse;
import static bouchonnois.tests.builders.PartieDeChasseBuilder.surUnTerrainRicheEnGalinettes;
import static bouchonnois.tests.builders.PartieDeChasseBuilder.unePartieDeChasseInexistante;
import static org.assertj.core.api.Assertions.assertThat;

class PrendreLApéro extends PartieDeChasseServiceTests {
    @Test
    void quand_la_partie_est_en_cours() {
        given(
                unePartieDeChasseExistante(
                        surUnTerrainRicheEnGalinettes()
                )
        );

        when(id -> partieDeChasseService.prendreLapéro(id));

        then(savedPartieDeChasse ->
                assertPartieDeChasse(savedPartieDeChasse)
                        .hasEmittedEvent(now, "Petit apéro")
                        .beInApéro()
        );
    }

    @Nested
    class Echoue {
        @Test
        void car_partie_nexiste_pas() {
            given(unePartieDeChasseInexistante());

            when(id -> partieDeChasseService.prendreLapéro(id));

            thenThrow(LaPartieDeChasseNexistePas.class, savedPartieDeChasse -> assertThat(savedPartieDeChasse).isNull());
        }

        @Test
        void si_les_chasseurs_sont_déjà_en_apéro() {
            given(unePartieDeChasseExistante(
                    surUnTerrainRicheEnGalinettes()
                            .aLapéro()
            ));

            when(id -> partieDeChasseService.prendreLapéro(id));

            thenThrow(OnEstDéjaEnTrainDePrendreLapéro.class, savedPartieDeChasse -> assertThat(savedPartieDeChasse).isNull());
        }

        @Test
        void si_la_partie_de_chasse_est_terminée() {
            given(unePartieDeChasseExistante(
                    surUnTerrainRicheEnGalinettes()
                            .terminée()
            ));

            when(id -> partieDeChasseService.prendreLapéro(id));

            thenThrow(OnPrendPasLapéroQuandLaPartieEstTerminée.class, savedPartieDeChasse -> assertThat(savedPartieDeChasse).isNull());
        }
    }
}